### PR TITLE
feat: rudimentary authentication

### PR DESCRIPTION
**Pull request by asherlloyd**
## Description
Resolves several items on issue [#29](https://github.com/PrayTeam/scriptured-prayer/issues/29).

## Reason
To provide authentication and basic user state management. Also resolves some CORS conundrums. There are still outstanding items to resolve on [#29](https://github.com/PrayTeam/scriptured-prayer/issues/29).

## How this has been tested
Tested Django build and React development server integration.

## Types of changes
- [x] New feature (adds functionality)
- [ ] Bug fix (fixes an issue)
- [ ] Breaking changes (causes defects to existing pieces)

## Screenshots
image
image

## Checklist:
- [x] I have read and complied with the **CONTRIBUTING** document
- [x] The changes I've made follow the existing code style and formatting requirements
- [ ] My changes require updates to the documentation, which I have modified accordingly

### CODE REVIEW

**Review by kenancasey — APPROVED** [review]
I'm good with this.

File: scriptured_prayer/settings.py [hunk 1]
```diff
@@ -26,7 +26,11 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
 
-ALLOWED_HOSTS = []
+# disable when going to prod
```
Comment (Soyokaze-42) [inline scriptured_prayer/settings.py:29]: This is actually what we are planning on using in prod, too.

File: scriptured_prayer/settings.py [hunk 2]
```diff
@@ -160,10 +165,12 @@
 }
 
 REST_FRAMEWORK = {
-    # Use Django's standard `django.contrib.auth` permissions,
-    # or allow read-only access for unauthenticated users.
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        'rest_framework.authentication.SessionAuthentication',
+    ),
```
Comment (Soyokaze-42) [inline scriptured_prayer/settings.py:170]: What are we gaining by adding IsAuthenticated here? It looks like you are overriding this setting in all the views you've added.

File: prayerapp/templates/prayerapp/index.html [hunk 3]
```diff
@@ -1,6 +1,8 @@
 {% load static %}
+{% load i18n %}
+{% get_current_language as LANGUAGE_CODE %}
 <!doctype html>
-<html lang="en">
+<html lang="{{ LANGUAGE_CODE }}">
```
Comment (Soyokaze-42) [inline prayerapp/templates/prayerapp/index.html:5]: This looks like an efficient way to get the lang code into js.

File: prayerapp/views.py [hunk 4]
```diff
@@ -157,4 +164,73 @@ class CardViewSet(viewsets.ReadOnlyModelViewSet):
...
+        return Card.objects.all().prefetch_related("category")
+
+@method_decorator(ensure_csrf_cookie, name='dispatch')
+class CSRFView(APIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def get(self, request, format=None):
+        return Response(True)
+
+@method_decorator(csrf_protect, name='dispatch')
```
Comment (Soyokaze-42) [inline prayerapp/views.py:176]: csrf_protect is enabled by default with middleware, so, unless this is an edge case I'm unaware of, this decorator is redundant.

File: prayerapp/views.py [hunk 5]
```diff
@@ -157,4 +164,73 @@ class CardViewSet(viewsets.ReadOnlyModelViewSet):
...
+    def post(self, request, format=None):
+        try:
+            logout(request)
+            return JsonResponse({ 'success': 'User logged out successfully' })
+        except:
+            return JsonResponse({ 'error': 'Failed to logout user' })
+
+
+@method_decorator(csrf_protect, name='dispatch')
+class RegisterView(APIView):
```
Comment (Soyokaze-42) [inline prayerapp/views.py:217]: This raises a NameError when run. User not defined.

File: prayerapp/views.py [hunk 6]
```diff
@@ -157,4 +164,73 @@ class CardViewSet(viewsets.ReadOnlyModelViewSet):
...
+        if username is None or password is None:
+            return JsonResponse({
+                "errors": {
+                    "__all__": "Please enter a valid username and password"
+                }
+            }, status=400)
+
+        user = authenticate(username=username, password=password)
+
+        if user is not None:
```
Comment (Soyokaze-42) [inline prayerapp/views.py:195]: Getting the user data should be done in a DRF serializer.

File: prayerapp/views.py [hunk 7]
```diff
@@ -157,4 +164,73 @@ class CardViewSet(viewsets.ReadOnlyModelViewSet):
     filterset_class = CardFilter
 
     def get_queryset(self):
-        return Card.objects.all().prefetch_related("category")
\ No newline at end of file
+        return Card.objects.all().prefetch_related("category")
+
+@method_decorator(ensure_csrf_cookie, name='dispatch')
+class CSRFView(APIView):
```
Comment (Soyokaze-42) [inline prayerapp/views.py:170]: What do we need this for? Is this just to check that there is a csrf token? If so, django should be checking that on every request without any configuration changes.

**Review by Soyokaze-42 — COMMENTED** [review]
Instead of rolling our own auth endpoints, what do you think about installing this instead: https://github.com/iMerica/dj-rest-auth? I haven't used it before, but it isn't dead. This works with session auth and the DRF token auth out of the box.

**Comment by Soyokaze-42** [comment]
I get an exception when I visit http://localhost:8000/en/api/schema/ in this branch atm

**Comment by Soyokaze-42** [comment]
It seems like there is an issue between drf-spectacular and df_rest_auth. We can either go back to writing our own, trash swagger, find a setting or code to fix the issue, or find another rest auth module. I read through the drf-spectacular docs, but nothing jumped out at me as a solution.

**Comment by Soyokaze-42** [comment]
@asherlloyd I found a fix for the drf-spectacular. There is a way to override the classes used to generate the schema. I'm not sure what it is replacing or adding, but it fixed the issue.

I also moved the auth api endpoint slightly (to /api/auth) so we can avoid potential future name clashes.

**Comment by Soyokaze-42** [comment]
I did not update the javascript, though, so I probably broke it :(

**Comment by asherlloyd** [comment]
> I did not update the javascript, though, so I probably broke it :(

No problem! JavaScript URLs updated to match.

**Pull request merged**